File: Backend/document-service/services/embeddingService.js
const crypto = require('crypto');
const { GoogleGenAI } = require('@google/genai');
const ChunkEmbeddingCache = require('../models/ChunkEmbeddingCache');
const embeddingRateLimiter = require('../utils/embeddingRateLimiter');

// Initialize the Gemini client with API key
const client = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

const MAX_CHARS = Number(process.env.GEMINI_EMBEDDING_MAX_CHARS || 10000);
const BATCH_SIZE = Number(process.env.GEMINI_EMBEDDING_BATCH_SIZE || 100);

// Use the correct Gemini embedding model
const PRIMARY_MODEL = process.env.GEMINI_EMBEDDING_MODEL || 'gemini-embedding-001';
const OUTPUT_DIMENSIONALITY = Number(process.env.GEMINI_EMBEDDING_DIMENSION || 768);

console.log(`[EmbeddingService] Initialized with model: ${PRIMARY_MODEL}`);
console.log(`[EmbeddingService] Output dimensionality: ${OUTPUT_DIMENSIONALITY}`);
console.log(`[EmbeddingService] Batch size: ${BATCH_SIZE}`);

const MAX_RETRIES = 3;
const RETRY_DELAY_BASE = 1000;

function cleanText(text) {
  if (!text) return '';
  return text.replace(/\s+/g, ' ').trim().slice(0, MAX_CHARS);
}

function computeContentHash(text) {
  return crypto.createHash('sha256').update(text || '', 'utf8').digest('hex');
}

/**
 * Normalize embeddings for dimensions other than 3072
 * This ensures accurate semantic similarity comparisons
 */
function normalizeEmbedding(embedding) {
  if (!Array.isArray(embedding) || embedding.length === 0) {
    throw new Error('Invalid embedding array');
  }

  // Calculate the magnitude (L2 norm)
  const magnitude = Math.sqrt(embedding.reduce((sum, val) => sum + val * val, 0));

  if (magnitude === 0) {
    throw new Error('Cannot normalize zero vector');
  }

  // Normalize each value
  return embedding.map(val => val / magnitude);
}

/**
 * Generate embeddings using the new Gemini API
 */
async function embedBatchWithModel(modelName, texts, retryCount = 0) {
  console.log(`[EmbeddingService] Embedding ${texts.length} texts with ${modelName}`);

  try {
    // Use rate limiter to ensure we don't exceed API limits
    const result = await embeddingRateLimiter.addRequest(async () => {
      return await client.models.embedContent({
        model: modelName,
        contents: texts.map(text => cleanText(text)),
        config: {
          taskType: 'RETRIEVAL_DOCUMENT', // Optimized for document search/RAG
          outputDimensionality: OUTPUT_DIMENSIONALITY
        }
      });
    });

    console.log(`[EmbeddingService] API Response received`);
    console.log(`[EmbeddingService] Result type: ${typeof result}`);
    console.log(`[EmbeddingService] Result keys: ${result ? Object.keys(result).join(', ') : 'null'}`);
    console.log(`[EmbeddingService] result.embeddings type: ${typeof result?.embeddings}`);
    console.log(`[EmbeddingService] result.embeddings length: ${result?.embeddings?.length}`);

    if (!result) {
      throw new Error('API returned null/undefined result');
    }

    if (!result.embeddings) {
      console.error(`[EmbeddingService] ❌ result.embeddings is undefined!`);
      console.error(`[EmbeddingService] Full result:`, JSON.stringify(result, null, 2));
      throw new Error('API response missing embeddings property');
    }

    if (!Array.isArray(result.embeddings)) {
      console.error(`[EmbeddingService] ❌ result.embeddings is not an array!`);
      console.error(`[EmbeddingService] Type: ${typeof result.embeddings}`);
      throw new Error(`API response embeddings is not an array: ${typeof result.embeddings}`);
    }

    if (result.embeddings.length === 0) {
      console.error(`[EmbeddingService] ❌ result.embeddings is empty array!`);
      throw new Error('API returned empty embeddings array');
    }

    console.log(`[EmbeddingService] ✅ Successfully embedded ${texts.length} texts`);

    // Extract embeddings and normalize if needed
    const embeddings = result.embeddings.map((embeddingObj, idx) => {
      if (!embeddingObj) {
        console.error(`[EmbeddingService] ❌ Embedding at index ${idx} is null/undefined`);
        throw new Error(`Embedding at index ${idx} is null/undefined`);
      }

      if (!embeddingObj.values) {
        console.error(`[EmbeddingService] ❌ Embedding at index ${idx} missing 'values' property`);
        console.error(`[EmbeddingService] Embedding keys: ${Object.keys(embeddingObj).join(', ')}`);
        throw new Error(`Embedding at index ${idx} missing 'values' property`);
      }

      const values = embeddingObj.values;

      if (!Array.isArray(values)) {
        console.error(`[EmbeddingService] ❌ Embedding values at index ${idx} is not an array`);
        throw new Error(`Embedding values at index ${idx} is not an array`);
      }

      if (values.length === 0) {
        console.error(`[EmbeddingService] ❌ Embedding values at index ${idx} is empty`);
        throw new Error(`Embedding values at index ${idx} is empty`);
      }

      // Normalize for dimensions other than 3072
      if (OUTPUT_DIMENSIONALITY !== 3072) {
        return normalizeEmbedding(values);
      }

      return values;
    });

    console.log(`[EmbeddingService] ✅ Extracted and processed ${embeddings.length} embeddings`);
    console.log(`[EmbeddingService] First embedding dimension: ${embeddings[0]?.length}`);

    return embeddings;
  } catch (error) {
    console.error(`[EmbeddingService] ❌ Error with model ${modelName}:`, error.message);
    console.error(`[EmbeddingService] Error stack:`, error.stack);

    const errorMessage = error?.message || String(error);
    const isRateLimitError = errorMessage.includes('429') ||
      errorMessage.includes('Too Many Requests') ||
      errorMessage.includes('quota') ||
      errorMessage.includes('rate limit');

    // Handle rate limit errors with exponential backoff
    if (isRateLimitError && retryCount < MAX_RETRIES) {
      const retryDelay = RETRY_DELAY_BASE * Math.pow(2, retryCount);
      console.warn(`[EmbeddingService] Rate limit hit. Retrying in ${retryDelay}ms (attempt ${retryCount + 1}/${MAX_RETRIES})`);

      await new Promise(resolve => setTimeout(resolve, retryDelay));
      return embedBatchWithModel(modelName, texts, retryCount + 1);
    }

    throw error;
  }
}

/**
 * Generate embeddings with metadata
 */
async function generateEmbeddingsWithMeta(texts) {
  if (!Array.isArray(texts) || texts.length === 0) {
    console.warn('[EmbeddingService] No texts provided for embedding');
    return { embeddings: [], model: PRIMARY_MODEL };
  }

  const results = [];
  const totalBatches = Math.ceil(texts.length / BATCH_SIZE);

  console.log(`[EmbeddingService] Processing ${texts.length} texts in ${totalBatches} batches`);

  // Process batches sequentially to stay within rate limits
  for (let batchIndex = 0; batchIndex < totalBatches; batchIndex++) {
    const batchStart = batchIndex * BATCH_SIZE;
    const batch = texts.slice(batchStart, batchStart + BATCH_SIZE);

    if (batch.length === 0) continue;

    try {
      const embeddings = await embedBatchWithModel(PRIMARY_MODEL, batch);
      results.push(...embeddings);

      // Log progress every 10 batches or on completion
      if ((batchIndex + 1) % 10 === 0 || batchIndex === totalBatches - 1) {
        const progress = ((batchIndex + 1) / totalBatches * 100).toFixed(1);
        console.log(`[EmbeddingService] ✅ Progress: ${batchIndex + 1}/${totalBatches} batches (${progress}%)`);
      }
    } catch (error) {
      console.error(`[EmbeddingService] ❌ Failed to embed batch ${batchIndex + 1}/${totalBatches}:`, error.message);
      throw new Error(`Embedding generation failed at batch ${batchIndex + 1}: ${error.message}`);
    }
  }

  console.log(`[EmbeddingService] ✅ Completed embedding ${texts.length} texts`);
  console.log(`[EmbeddingService] Total embeddings generated: ${results.length}`);
  console.log(`[EmbeddingService] Embedding dimension: ${results[0]?.length || 'N/A'}`);

  return { embeddings: results, model: PRIMARY_MODEL };
}

/**
 * Generate a single embedding
 */
async function generateEmbedding(text) {
  if (!text || typeof text !== 'string') {
    throw new Error('Invalid text provided for embedding');
  }

  const { embeddings } = await generateEmbeddingsWithMeta([text]);

  if (!embeddings || embeddings.length === 0) {
    throw new Error('Failed to generate embedding');
  }

  return embeddings[0];
}

/**
 * Generate multiple embeddings
 */
async function generateEmbeddings(texts) {
  if (!Array.isArray(texts)) {
    throw new Error('texts must be an array');
  }

  const { embeddings } = await generateEmbeddingsWithMeta(texts);

  if (embeddings.length !== texts.length) {
    throw new Error(`Embedding count mismatch: expected ${texts.length}, got ${embeddings.length}`);
  }

  return embeddings;
}

/**
 * Get cached embedding by content hash
 */
async function getCachedEmbedding(hash) {
  if (!hash) return null;
  try {
    return await ChunkEmbeddingCache.getEmbeddingByHash(hash);
  } catch (error) {
    console.warn('[EmbeddingService] Cache lookup failed:', error?.message || error);
    return null;
  }
}

/**
 * Cache an embedding
 */
async function cacheEmbedding({ hash, embedding, model, tokenCount }) {
  try {
    await ChunkEmbeddingCache.upsertEmbedding({
      contentHash: hash,
      embedding,
      model,
      tokenCount,
    });
  } catch (error) {
    console.warn('[EmbeddingService] Cache upsert failed:', error?.message || error);
  }
}

module.exports = {
  cleanText,
  computeContentHash,
  generateEmbedding,
  generateEmbeddings,
  generateEmbeddingsWithMeta,
  normalizeEmbedding,
  getCachedEmbedding,
  cacheEmbedding,
  BATCH_SIZE,
  PARALLEL_BATCHES: Number(process.env.EMBEDDING_PARALLEL_BATCHES || 3),
  embeddingRateLimiter,
};

File: Backend/agent-draft-service/services/embedding_service.py
# """Gemini embedding service. Mirrors document-service embeddingService (models/text-embedding-004)."""

# from __future__ import annotations

# import os
# from typing import List

# from google import genai  # type: ignore


# MAX_CHARS = int(os.environ.get("GEMINI_EMBEDDING_MAX_CHARS", "8000"))
# BATCH_SIZE = int(os.environ.get("GEMINI_EMBEDDING_BATCH_SIZE", "100"))
# MODEL = os.environ.get("GEMINI_EMBEDDING_MODEL", "models/text-embedding-004")


# def _client() -> genai.Client:
#     api_key = os.environ.get("GEMINI_API_KEY") or os.environ.get("GOOGLE_API_KEY")
#     if not api_key:
#         raise ValueError("GEMINI_API_KEY or GOOGLE_API_KEY must be set")
#     return genai.Client(api_key=api_key)


# def _clean_text(text: str) -> str:
#     if not text:
#         return ""
#     return " ".join(text.split()).strip()[:MAX_CHARS]


# def generate_embeddings(texts: List[str]) -> List[List[float]]:
#     """
#     Generate embeddings for a list of texts using Gemini models/text-embedding-004.
#     Mirrors document-service embeddingService.generateEmbeddings.
#     """
#     import logging
#     import time
#     logger = logging.getLogger(__name__)
    
#     if not texts:
#         return []

#     client = _client()
#     cleaned = [_clean_text(t) for t in texts]
#     all_embeddings: List[List[float]] = []
    
#     max_retries = 3
#     retry_delay = 1  # seconds

#     for i in range(0, len(cleaned), BATCH_SIZE):
#         batch = cleaned[i : i + BATCH_SIZE]
        
#         for attempt in range(max_retries):
#             try:
#                 response = client.models.embed_content(
#                     model=MODEL,
#                     contents=batch,
#                 )
#                 if hasattr(response, "embeddings") and response.embeddings:
#                     for emb in response.embeddings:
#                         vals = getattr(emb, "values", None) or []
#                         all_embeddings.append(list(vals))
#                 else:
#                     # Single content: response.embedding or response.embeddings[0]
#                     emb = getattr(response, "embedding", None)
#                     if emb is None and getattr(response, "embeddings", None):
#                         emb = response.embeddings[0]
#                     if emb is not None:
#                         v = getattr(emb, "values", None)
#                         all_embeddings.append(list(v) if v is not None else [])
#                 break  # Success, exit retry loop
                
#             except Exception as e:
#                 error_msg = str(e)
#                 if attempt < max_retries - 1:
#                     logger.warning(f"[Embedding Service] Attempt {attempt + 1}/{max_retries} failed: {error_msg}. Retrying in {retry_delay}s...")
#                     time.sleep(retry_delay)
#                     retry_delay *= 2  # Exponential backoff
#                 else:
#                     logger.error(f"[Embedding Service] All {max_retries} attempts failed: {error_msg}")
#                     raise

#     return all_embeddings
"""Gemini embedding service using gemini-embedding-001."""

from __future__ import annotations

import os
from typing import List

import numpy as np
from google import genai  # type: ignore
from google.genai import types


MAX_CHARS = int(os.environ.get("GEMINI_EMBEDDING_MAX_CHARS", "8000"))
BATCH_SIZE = int(os.environ.get("GEMINI_EMBEDDING_BATCH_SIZE", "100"))
# Default to gemini-embedding-001 (official Gemini embedding model)
MODEL = os.environ.get("GEMINI_EMBEDDING_MODEL", "gemini-embedding-001")
# Output dimensions: 768, 1536, or 3072 (default 768 to match database schema)
EMBEDDING_DIMENSIONS = int(os.environ.get("GEMINI_EMBEDDING_DIMENSIONS", "768"))

# Note: gemini-embedding-001 doesn't require 'models/' prefix


def _client() -> genai.Client:
    api_key = os.environ.get("GEMINI_API_KEY") or os.environ.get("GOOGLE_API_KEY")
    if not api_key:
        raise ValueError("GEMINI_API_KEY or GOOGLE_API_KEY must be set")
    return genai.Client(api_key=api_key)


def _clean_text(text: str) -> str:
    if not text:
        return ""
    return " ".join(text.split()).strip()[:MAX_CHARS]


def generate_embeddings(texts: List[str]) -> List[List[float]]:
    """
    Generate embeddings for a list of texts using Gemini gemini-embedding-001.
    """
    import logging
    import time
    logger = logging.getLogger(__name__)
    
    if not texts:
        return []

    client = _client()
    cleaned = [_clean_text(t) for t in texts]
    all_embeddings: List[List[float]] = []
    
    max_retries = 3
    retry_delay = 1  # seconds

    for i in range(0, len(cleaned), BATCH_SIZE):
        batch = cleaned[i : i + BATCH_SIZE]
        
        for attempt in range(max_retries):
            try:
                logger.info(f"[Embedding Service] Processing batch {i//BATCH_SIZE + 1}, size={len(batch)}, model={MODEL}, dimensions={EMBEDDING_DIMENSIONS}")
                
                response = client.models.embed_content(
                    model=MODEL,
                    contents=batch,
                    config=types.EmbedContentConfig(
                        output_dimensionality=EMBEDDING_DIMENSIONS
                    )
                )
                
                if hasattr(response, "embeddings") and response.embeddings:
                    for emb in response.embeddings:
                        vals = getattr(emb, "values", None) or []
                        vals_list = list(vals)
                        
                        # Normalize embeddings for dimensions other than 3072
                        # (3072 is already normalized by the model)
                        if EMBEDDING_DIMENSIONS != 3072 and vals_list:
                            vals_np = np.array(vals_list)
                            norm = np.linalg.norm(vals_np)
                            if norm > 0:
                                vals_list = (vals_np / norm).tolist()
                        
                        all_embeddings.append(vals_list)
                else:
                    # Single content: response.embedding or response.embeddings[0]
                    emb = getattr(response, "embedding", None)
                    if emb is None and getattr(response, "embeddings", None):
                        emb = response.embeddings[0]
                    if emb is not None:
                        v = getattr(emb, "values", None)
                        vals_list = list(v) if v is not None else []
                        
                        # Normalize embeddings for dimensions other than 3072
                        if EMBEDDING_DIMENSIONS != 3072 and vals_list:
                            vals_np = np.array(vals_list)
                            norm = np.linalg.norm(vals_np)
                            if norm > 0:
                                vals_list = (vals_np / norm).tolist()
                        
                        all_embeddings.append(vals_list)
                
                logger.info(f"[Embedding Service] Successfully generated {len(batch)} embeddings")
                break  # Success, exit retry loop
                
            except Exception as e:
                error_msg = str(e)
                if attempt < max_retries - 1:
                    logger.warning(f"[Embedding Service] Attempt {attempt + 1}/{max_retries} failed: {error_msg}. Retrying in {retry_delay}s...")
                    time.sleep(retry_delay)
                    retry_delay *= 2  # Exponential backoff
                else:
                    logger.error(f"[Embedding Service] All {max_retries} attempts failed: {error_msg}")
                    raise

    logger.info(f"[Embedding Service] Total embeddings generated: {len(all_embeddings)}")
    return all_embeddings